Fixes hostile mobs attacking surrounding tiles when trying to attack someone: the stunning finale

GupGup opened the pull request:
sorry for all these dumb prs, im new to git

fix for issue #4237
replacement for #32593 and #32665
revises #30152 (useless)

:cl: GupGup
fix: Fixes hostile mobs attacking surrounding tiles when trying to attack someone
/:cl:

GupGup committed:
fix for issue #4237
replacement for #32593 and #32665
revises #30152 (useless)

renamed DestroySurroundings proc to DestroyPathToTarget to better illustrate its purpose
kept original DestroySurroundings proc for use with megafauna since its other implied purpose is still necessary

wrote DestroyPathToTarget proc (tested)
it finds the optimal path to the target and least optimal path to the target, attempts to break rwalls, walls, and windows in that order, and if it cant, checks the surrounding turfs until it can (using the optimal path first and least optimal path last)

got rid of CanSmashTurfs proc (redundant and useless) and got rid of it being called in the code for the gorilla mob

added isreinforcedwallturf helper for use with DestroyPathToTarget

added get_cardinal_step_towards and get_next_cardinal_dir procs for use with DestroyPathToTarget

renamed DestroySurroundings to DestroyPathToTarget in the mimic and statue codes to use the correct functionality

ChangelingRain commented on code/__HELPERS/game.dm:
```diff
@@ -300,6 +300,44 @@
+			return get_step(start, EAST)
+
+
+/proc/get_next_cardinal_dir(original_direction, clockwise) // get the next cardinal direction clockwise from the current cardinal direction
```
this proc is covered by the existing byond function of `turn(dir, angle)`

ChangelingRain commented on code/modules/mob/living/simple_animal/hostile/hostile.dm:
```diff
@@ -359,24 +361,60 @@
 		P.fire()
 		return P
 
-/mob/living/simple_animal/hostile/proc/CanSmashTurfs(turf/T)
```
this should not be removed.

GupGup replied:
why not

GupGup replied:
by why, i mean the same functionality is already provided by environment_smash
it doesnt even have a real purpose and is only called by destroysurroundings and gorilla when the environment_smash thing should probably be used instead
tell me  if im wrong because im new to the codebase, its just how i see it

ChangelingRain replied:
It lets you control what walls a mob can smash if it's able to smash walls, which is important.

GupGup replied:
again, not trying to argue for arguments sake or anything, but its not really used right now and theres no real reason itll ever be used in the future (as far as i know)
i dont really see a difference in the types of walls we have so the difference that this proc talks about seems to be irrelevent

GupGup replied:
also, environment_smash is never called aside from this proc and your suggested code doesnt even use it, so it seems superfluous to have a bunch of environment_smash variables scattered around the code and a cansmashturfs proc

ChangelingRain commented on code/modules/mob/living/simple_animal/hostile/hostile.dm:
```diff
@@ -359,24 +361,60 @@
+			if(is_type_in_typecache(A, environment_target_typecache) && !A.IsObscured())
+				A.attack_animal(src)
+				dir_list.len = 0
+	del(dir_list)
```
byond garbage collects.

ChangelingRain commented on code/modules/mob/living/simple_animal/hostile/hostile.dm:
```diff
@@ -359,24 +361,60 @@
+
+	var/list/dir_list = list(dir_from_target, get_next_cardinal_dir(dir_from_target, 1), get_next_cardinal_dir(dir_to_target, 1), dir_to_target)
+
+	for(dir_list.len, dir_list.len > 0, dir_list.len--)
```
are the faster loop formats in contributor.md?
in any case, this loop is very odd and you can probably do it faster and find the dirs you need to attack much more easily.

GupGup replied:
i cant find contributor.md
i also dont know much about optimization, but if there are predefined loop formats, id be happy to try to apply them

GupGup replied:
nvm found it, its contributing.md

GupGup committed:
rewrote DestroyStructuresCycle to be faster and more efficient
also got rid of get_next_cardinal_dir

ChangelingRain commented on code/modules/mob/living/simple_animal/hostile/hostile.dm:
```diff
@@ -81,7 +81,9 @@
 	if(AICanContinue(possible_targets))
-		DestroySurroundings()
+		if(!QDELETED(target)) // fix a runtime error, probably hacky
+			if(!targets_from.Adjacent(target)) // so the mob doesnt try to make a path when it can just hit the target
```
these two ifs can be combined into one; `if(!QDELETED(target) && !targets_from.Adjacent(target))`

GupGup replied:
can i do the same to these lines
if (iswallturf(T) || ismineralturf(T))
if (!isreinforcedwallturf(T))

ChangelingRain commented on code/modules/mob/living/simple_animal/hostile/hostile.dm:
```diff
@@ -359,24 +361,56 @@
+				var/atom/A = a
+				if(is_type_in_typecache(A, environment_target_typecache) && !A.IsObscured())
+					A.attack_animal(src)
+					break cycle_loop
```
you could just return instead of using a named loop.
also, I'd do this like the following, which is probably clearer and more like what I'm expecting the code to do;
````
	var/dir_to_target = get_cardinal_step_towards(targets_from, target)
	var/dir_list = list()
	if(dir_to_target in GLOB.diagonals) //it's diagonal, so we need two directions to hit
		for(var/direction in GLOB.cardinals)
			if(direction & dir_to_target)
				dir_list += direction
	else
		dir_list += dir_to_target
	var/turf/T
	for(var/direction in dir_list) //now we hit all of the directions we got in this fashion, since it's the only directions we should actually need
		T = get_step(targets_from, direction)
		if(CanSmashTurfs(T))
			T.attack_animal(src)
		for(var/a in T)
			var/atom/A = a
			if(is_type_in_typecache(A, environment_target_typecache) && !A.IsObscured())
				A.attack_animal(src) //we neither break nor return here, so that it hits both directions at once instead of possibly being inefficient. this is still better than hitting every possible turf at once, since it'll be aimed at the target.
````

GupGup replied:
i know its really a subjective choice in the end but i think its more realistic that a hostile mob would only attack one turf at a time
while hitting 2 turfs at a time can fix the diagonal issue cleanly and stuff, it doesnt make sense since it would be the only proc that allows mobs to do that consistently

ChangelingRain replied:
hitting both is cleaner in the case where one is easier to destroy than the other, but would be attacked second, and is thus better.

GupGup replied:
couldnt you just check to see which of the 2 is easier to destroy instead of hitting them both

ChangelingRain replied:
that's difficult to determine.

GupGup replied:
couldnt you just make them hit the walls first then hit the structures with lower integrity first

ChangelingRain commented on code/modules/mob/living/simple_animal/hostile/hostile.dm:
```diff
@@ -359,24 +361,56 @@
+	if(environment_smash)
+		EscapeConfinement()
+		var/turf/T = get_cardinal_step_towards(targets_from, target)
+		if(T.Adjacent(targets_from))
```
this if chain is... not good, and you don't actually need it

GupGup replied:
i agree that escape confinement isnt really necessary but i thought it might be used in some edge case
same probably goes for environment smash
and i didnt think that Adjacent was necessary since the proc only gets 1 step away, but i figured id leave it in there
if its all not needed, ill get rid of it

ChangelingRain replied:
EscapeConfinement() is needed, but the if chain to check wall types is really bad and should be handled by the walls themselves.
with my above code, this proc should be something like this;
````
	if(environment_smash)
		EscapeConfinement()
		DestroyStructuresCycle()
````

GupGup committed:
combined a couple if statements
replaced a labelled loop with return
got rid of a pesky if chain

ChangelingRain commented on code/modules/mob/living/simple_animal/hostile/hostile.dm:
```diff
@@ -365,35 +364,29 @@
 
 /mob/living/simple_animal/hostile/proc/DestroyPathToTarget()
-	if(environment_smash)
-		EscapeConfinement()
```
I said the if chain wasn't good, not that this wasn't.
this code is all that this should actually need to work.
````
	if(environment_smash)
		EscapeConfinement()
		DestroyStructuresCycle()
```

GupGup replied:
yea i reverted it in my working code
i didnt understand what u meant by if chain

ChangelingRain replied:
It's a big chain of if() else statements that other code already handles(or should handle)

ChangelingRain commented on code/modules/mob/living/simple_animal/hostile/hostile.dm:
```diff
@@ -365,35 +364,29 @@
 /mob/living/simple_animal/hostile/proc/DestroyStructuresCycle(turf/T)
```
You don't use this as a proc argument; it just obfuscates the code.

GupGup committed:
reverted change to gorilla.dm
got rid of destroystructurecycle proc
used changelingrains code for destroypathtotarget

GupGup committed:
removed isreinforcedwallturf check

not needed anymore

ChangelingRain commented on code/modules/mob/living/simple_animal/hostile/hostile.dm:
```diff
@@ -361,32 +361,34 @@
+	if(environment_smash)
+		EscapeConfinement()
+
+		var/dir_to_target_cardinal = get_dir(targets_from, get_cardinal_step_towards(targets_from, target))
```
you don't need this at all, since if `if(dir_to_target in GLOB.diagonals)` fails we know the target is either on your turf or in a cardinal direction, so we can add it to the list without checking it.

ChangelingRain commented on code/modules/mob/living/simple_animal/hostile/hostile.dm:
```diff
@@ -359,24 +360,52 @@
+	if(environment_smash)
+		EscapeConfinement()
+
+		var/dir_to_target_cardinal = get_dir(targets_from, get_cardinal_step_towards(targets_from, target))
```
you don't need this at all, since if `if(dir_to_target in GLOB.diagonals)` fails we know the target is either on your turf or in a cardinal direction, so we can add it to the list without checking it.

ChangelingRain commented on code/modules/mob/living/simple_animal/hostile/hostile.dm:
```diff
@@ -81,7 +81,8 @@
 
 	if(AICanContinue(possible_targets))
-		DestroySurroundings()
+		if(!QDELETED(target) && !targets_from.Adjacent(target))
```
and, since we know that the target both exists and isn't adjacent to us, we know that it'll be a direction and is thus safe to jam into the list without worrying about it.

ChangelingRain commented on code/modules/mob/living/simple_animal/hostile/hostile.dm:
```diff
@@ -359,24 +360,52 @@
 		EscapeConfinement()
 		for(var/dir in GLOB.cardinals)
 			var/turf/T = get_step(targets_from, dir)
-			if(CanSmashTurfs(T))
```
should probably keep this, but change this code to be this... and make it a proc, shared between this and DestroyPathToTarget(), to reduce code duplication and make changing it easier.
````
/mob/living/simple_animal/hostile/proc/DestroyObjectsInDirection(direction)
 	var/turf/T = get_step(targets_from, direction)
 	if(T.Adjacent(targets_from))
 		if(CanSmashTurfs(T))
 			T.attack_animal(src)
 		for(var/a in T)
 			var/atom/A = a
 			if(!A.Adjacent(targets_from))
 				continue
 			if(is_type_in_typecache(A, environment_target_typecache) && !A.IsObscured())
 				A.attack_animal(src)
````

GupGup replied:
can you put that code in, ive lost p much all will to keep going with this bug
been very mentally frustrating, but overall a rewarding learning experience i guess
its like you think youve done it and realize that you need to reformat everything youve written, add a couple procs, remove those procs, go back to the source
maybe im not fit to be a programmer

GupGup replied:
ill fix it tomorrow

GupGup committed:
got rid of get_cardinal_step_towards

GupGup committed:
removed dir_to_target_cardinal

GupGup committed:
added destroyobjectsindirection proc
rewrote destroysurroundings and destroypathtotarget to include it
dadada
